RFC: Add psoc6

nazrchorn opened the pull request:
Added Cypress PSoC6 platform

carlescufi commented:
@nazrchorn can you please squash your commits into logical units? more info here: http://docs.zephyrproject.org/contribute/contribute_guidelines.html

galak commented:
> @nazrchorn can you please squash your commits into logical units? more info here: http://docs.zephyrproject.org/contribute/contribute_guidelines.html

I asked @nazrchorn to push this so I could give him feedback on such things

dbkinder commented on samples/security/spm/spm_nspe/README.rst:
```diff
@@ -0,0 +1,47 @@
+.. _spm_psoc6:
+
+SPM base NS environment example for PSoC6  CM4
+###########
```
Title underline needs to be at least as long as the title

dbkinder commented on samples/security/spm/spm_nspe/README.rst:
```diff
@@ -0,0 +1,47 @@
+
+Overview
+********
+A simple SPM base NS environment example for PSoC6  CM4.
```
Use a complete sentence, ``This sample is a simple SPM...``

dbkinder commented on samples/security/spm/spm_nspe/README.rst:
```diff
@@ -0,0 +1,47 @@
+Needs Secure part on CM0+
+
+CM4(spm_nspe) sends debug log to Uart Port5 which is available through USB connection with the PSoC6 kit.
+CM0+(spm_spe) send debug messages to UART port 12(pin0(RX)/pin1(TX)) � connect external UART to UASB convertor to these pins on kit 
```
This file contains non-ACSII/non-UTF-8 characters (that `–`).   Please use an editor to replace  window-1252 characters with their ASCII equivalents.  (Windows editors are notorious for changing things line single ``'`` and double quotes ``"`` into "smart" quotes.)  You can use the linux comand ``file --mime-encoding | grep "unknown"`` to find these files.

dbkinder replied:
@d3zd3z Here's the non-UTF8 encoding problem (an en-dash)

OlegKapshii replied:
Already fixed. It seems you have an older version of the tree

dbkinder commented on samples/security/spm/spm_nspe/README.rst:
```diff
@@ -0,0 +1,47 @@
+CM0+(spm_spe) send debug messages to UART port 12(pin0(RX)/pin1(TX)) � connect external UART to UASB convertor to these pins on kit 
+
+Project description:
+- CM0p starts CM4 core
```
need a blank line before the first item of a list

dbkinder commented on samples/security/spm/spm_nspe/README.rst:
```diff
@@ -0,0 +1,47 @@
+
+Project description:
+- CM0p starts CM4 core
+- CM4 writes the message with the counter in loop to the shared memory block, triggers interrupt to CM0p core and wait on response.
```
please keep line length < 80 characters

dbkinder commented on samples/security/spm/spm_nspe/README.rst:
```diff
@@ -0,0 +1,47 @@
+
+Building and Running
+********************
+cmake -G"Eclipse CDT4 - Ninja" -DBOARD=cy8ckit_062_wifi_bt_m0 ..
```
script and code examples need to be included in a code-block directive with the body of the code block indented by three spaces:

    .. code-block: bash

       cmake ...

dbkinder commented on samples/security/spm/spm_nspe/README.rst:
```diff
@@ -0,0 +1,47 @@
+Sample Output
+=============
+
+***** Booting Zephyr OS v1.12.0-rc2-48-g2c3c674aa *****
```
Console output should also use a ``.. code-block`` directive:

    .. code-block: console

       ***** Booting Zephyr OS v1.12.0-rc2-48-g2c3c674aa *****
       CM4: Main

dbkinder commented on samples/security/spm/spm_nspe/README.rst:
```diff
@@ -0,0 +1,47 @@
+CM4: ipc_interrupt_handler - 1 INTR = 0x1000000
+CM4: ipc_interrupt_handler - 1 INTR = 0x1000000
+CM4: ipc_interrupt_handler - 1 INTR = 0x1000000
+�
```
change this non-UTF8 character into three periods ``...``  (If you're using a windows editor you should look at configuring it to only insert ASCII or UTF-8 characters.)

OlegKapshii commented:
Pushed requested changes

nazrchorn committed:
ext: hal: cypress: Import Cypress PDL for PSoC6 SoC Support

This includes:
* the Cypress Peripheral Driver Library (PDL) For PSoC6 SoC.

Origin: Cypress PDL for PSoC6
URL: https://github.com/cypress/pdl
Version: 3.1.0 (SHA: 2af49fe057320981c11117462a0962d956fc3056)
Purpose: Add support for Cypress PSoC SoC
License: Apache
Maintained-by: External

Signed-off-by: Oleg Kapshii <oleg.kapshii@cypress.com>
Signed-off-by: Nazar Chornenkyy <nazar.chornenkyy@cypress.com>

nazrchorn committed:
drivers: serial: Add Cypress PSoC6 UART driver

Added basic PSoC6 UART driver and added two UART nodes in the PSoC6
device tree to have output from CM0+ and CM4 cores.

Signed-off-by: Nazar Chornenkyy <nazar.chornenkyy@cypress.com>
Signed-off-by: Oleg Kapshii <oleg.kapshii@cypress.com>

nazrchorn committed:
boards: arm: Add Cypress PSoC6 cy8ckit_062_wifi_bt_m4 board support

Added cy8ckit_062_wifi_bt_m4 board configuration to build FW for CM4
core.
CM4 core does not start itself, it have to be started by CM0+ FW.

Signed-off-by: Nazar Chornenkyy <nazar.chornenkyy@cypress.com>
Signed-off-by: Oleg Kapshii <oleg.kapshii@cypress.com>

nazrchorn committed:
arm: Add Cypress PSoC6 SoC support

Added initial support and created the corresponding device tree part for
building PSoC6 SoC as part of Zephyr.

Signed-off-by: Nazar Chornenkyy <nazar.chornenkyy@cypress.com>
Signed-off-by: Oleg Kapshii <oleg.kapshii@cypress.com>

galak committed:
ext: hal: cypress: Protect floating point code with CONFIG_FLOAT

There is a little bit of floating point code in cy_sysclk.c.  For now
add compile protection for this code using CONFIG_FLOAT.  This code
utilizes some libm functions like ceil().  So if / when the functions
Cy_SysClk_EcoConfigure & Cy_SysClk_FllConfigure are needed we'll also
new to probably pull in newlib.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

nazrchorn committed:
ext: hal: cypress: Integrate Cypress PDL into Zephyr Build system

Add initial support for building Cypress PDL as part of Zephyr.

Signed-off-by: Nazar Chornenkyy <nazar.chornenkyy@cypress.com>
Signed-off-by: Oleg Kapshii <oleg.kapshii@cypress.com>

nazrchorn committed:
boards: arm: Add Cypress PSoC6 cy8ckit_062_wifi_bt_m0 board support

Added cy8ckit_062_wifi_bt_m0 board configuration to build FW for CM0+
core.

Signed-off-by: Nazar Chornenkyy <nazar.chornenkyy@cypress.com>
Signed-off-by: Oleg Kapshii <oleg.kapshii@cypress.com>

galak commented on drivers/serial/Kconfig.psoc6:
```diff
@@ -0,0 +1,35 @@
+
+if UART_PSOC6
+
+# --- port 6 ---
```
remove # --- port 6 --- line.

galak commented on drivers/serial/Kconfig.psoc6:
```diff
@@ -0,0 +1,35 @@
+
+# --- port 6 ---
+
+config UART_PSOC6_PORT_12
```
Best to sort these in numerical order, so PORT_5 first, than PORT_12

galak commented on drivers/serial/uart_psoc6.c:
```diff
@@ -0,0 +1,179 @@
+{
+	uint32_t rec;
+
+	//Uart *const uart = DEV_CFG(dev)->regs;
```
Remove unused code.

galak commented on drivers/serial/uart_psoc6.c:
```diff
@@ -0,0 +1,179 @@
+ * Note:
+ * - Error handling is not implemented.
+ * - The driver works only in polling mode, interrupt mode is not implemented.
+ * - Only port 12 is supported yet
```
Fix comment about only port 12 supported, both 5 & 12 are supported.

galak reviewed (changes requested):
* The driver needs cleaning up to support having multiple UARTs enabled at one time.
* Driver should use device tree for UART params like UART_SCB (ie the base address).  See dts/bindings/serial for examples.
* Fix inline comments.

galak commented on arch/arm/soc/cypress/psoc6/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,42 @@
+	# must be >= the highest interrupt number used
+	default 40
+
+	config SYS_CLOCK_HW_CYCLES_PER_SEC
```
fix whitespace

galak commented on arch/arm/soc/cypress/psoc6/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,42 @@
+
+if SOC_PSOC6_M0
+
+config UART_PSOC6_PORT_12
```
this seems like a board config choice, not SoC.

galak commented on arch/arm/soc/cypress/psoc6/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,42 @@
+
+if SOC_PSOC6_M4
+
+config UART_PSOC6_PORT_5
```
Seems like should be board, not SoC doing this.

galak replied:
I made a number of fixes to the SoC code

galak commented:
@OlegKapshii Can you focus on the comments related to the uart driver.

OlegKapshii commented:
I believe, that the PSoC6 UART driver is ready.

OlegKapshii commented:
Not sure what should I do more here.

pfalcon commented:
@OlegKapshii 

> Not sure what should I do more here.

Please see this comment https://github.com/zephyrproject-rtos/zephyr/pull/8408#issuecomment-414376131 for specific things to address, for starters. Beyond that, please don't include merge commits in PRs, rebase instead (so, this PR needs to be rebased). Thanks.

pfalcon commented on arch/arm/soc/cypress/psoc6/linker.ld:
```diff
@@ -0,0 +1,16 @@
+/*
+ * Copyright (c) 2014 Wind River Systems, Inc.
```
Please use up to date copyright here (Cypress or Linaro).

pfalcon commented on dts/arm/cypress/psoc6.dtsi:
```diff
@@ -0,0 +1,83 @@
+
+	soc {
+		uart5: uart@40001000 {
+			compatible = "cypress,nrf-uart";
```
The name here should be "cypress,psoc6-uart". (Unless you know that it's the same UART as in psoc5, psoc4, etc., in which case it can be "cypress,psoc-uart".)

OlegKapshii replied:
Already fixed in "Refactored Cypress PSoC6 UART driver" commit

pfalcon replied:
> Already fixed in "Refactored Cypress PSoC6 UART driver" commit

Yeah, I figured that later too. That's why, as said in https://github.com/zephyrproject-rtos/zephyr/pull/8408#issuecomment-397592114 and https://github.com/zephyrproject-rtos/zephyr/pull/8408#issuecomment-414376131 , it's important to squash fixup into the main commit, so people looking at a particular commit, so the latest, clean code.

Because otherwise, just consider yourself: this PR changes/adds 304 files. It's very hard to review it all at once. Even knowing that there's e.g. vendor HAL, which should be skipped, it's hard to find files which should be reviewed, and which not.

The only realistic way is to review commit by commit, and then it's important that fix-ups are squashed into each separate commits they affect.

pfalcon commented on dts/arm/cypress/psoc6_cm0.dtsi:
```diff
@@ -0,0 +1,28 @@
+
+/ {
+	cpus {
+		cpu@0 {
```
Why repeat here what's in included psoc6.dtsi?

OlegKapshii replied:
Made by analogy to dts\arm\nxp\nxp_lpc54xxx.dtsi, nxp_lpc54xxx_m0.dtsi, nxp_lpc54xxx_m4.dtsi

pfalcon commented on dts/arm/cypress/psoc6_cm0.dtsi:
```diff
@@ -0,0 +1,28 @@
+	soc {
+		/delete-node/ uart5;
+
+		uart12: uart@40002000 {
```
Why repeat here what's in included psoc6.dtsi?

pfalcon commented on dts/arm/cypress/psoc6_cm4.dtsi:
```diff
@@ -0,0 +1,28 @@
+	cpus {
+		/delete-node/ cpu@0;
+
+		cpu@1 {
```
Why repeat here what's in included psoc6.dtsi?

pfalcon commented on dts/arm/cypress/psoc6_cm4.dtsi:
```diff
@@ -0,0 +1,28 @@
+	};
+
+	soc {
+		uart5: uart@40001000 {
```
Why repeat here what's in included psoc6.dtsi?

pfalcon commented on drivers/serial/uart_psoc6.c:
```diff
@@ -28,57 +27,33 @@
-/* Allocate context for UART operation */
-cy_stc_scb_uart_context_t uartContext;
+struct cypress_psoc6_config {
+	CySCB_Type *base;
```
Inconsistent indentation.

pfalcon replied:
More specifically, should use tabs for indentation everywhere.

pfalcon commented on boards/arm/cy8ckit_062_wifi_bt_m4/doc/cy8ckit_062_wifi_bt.rst:
```diff
@@ -0,0 +1,115 @@
+platform that enables design and debug of the PSoC 62 MCU and the Murata
+LBEE5KL1DX Module (CYW4343W WiFi + Bluetooth Combo Chip).
+The PSoC 6 WiFi-BT Pioneer Kit features the PSoC 62 MCU: a single- or
+dual-core MCU, with an Arm Cortex-M4 and Arm Cortex-M0+, 1MB of Flash, 288KB
```
We should describe a specific board here, and as the CY8CKIT-WIFI-BT board has dual-core CPU, it should say that just.

pfalcon commented on boards/arm/cy8ckit_062_wifi_bt_m4/doc/cy8ckit_062_wifi_bt.rst:
```diff
@@ -0,0 +1,115 @@
++-----------+------------+----------------------+
+| UART      | on-chip    | serial port          |
++-----------+------------+----------------------+
+| GPIO      | on-chip    | gpio                 |
```
As far as I can tell, GPIO is not supported by this PR.

pfalcon commented:
@nazrchorn, @OlegKapshii: It doesn't end there though. I can't get the serial output for BOARD=cy8ckit_062_wifi_bt_m0 either. Can you confirm that P12_0, P12_1 pins should be used in this case? Because when I attach the USB-UART adapter you gave me (based Cypress chip) to these pins, I don't get anything (flashed with the command above). To be sure, I swapped wires to the pins too.

But I have a feeling that the problem can be also with this adapter's driver/support in Linux.

Bottom line: again, need the exact docs for serial access and flashing to avoid guesswork and confusion.

@galak : FYI

OlegKapshii commented:
Yes, BOARD=cy8ckit_062_wifi_bt_m0 works with P12 port, which is translated to USB socket.
Actually you should work with cy8ckit_062_wifi_bt_m0 for hello_world example. CM4 core does not starts by default, only CM0+ core works.
In order to have CM4 core working you should write FW for both cores (the same hello_world example, but with an additional command in main for CM0+ core) and CM0+ FW should start CM4 core using command in main - Cy_SysEnableCM4(0x10060000);
In any case, you should see the CM0+ output with BOARD=cy8ckit_062_wifi_bt_m0.

pfalcon commented:
@OlegKapshii: Thanks for quick response!

> Yes, BOARD=cy8ckit_062_wifi_bt_m0 works with P12 port, which is translated to USB socket.

Either I misunderstood what you wrote, or that's not correct, according to CY8CKIT-062-WiFi-BT_Kit_Guide.pdf, section 4.3.4, page 54, which says that PSoC6's pins P5[0], P5[1], P5[2], P5[3] are converted to onboard KitProg2's USB-UART Bridge. I.e. M4's UART on P5 would be exposed in this way, not M0's on P12. I know it's easy to mix up such things (especially if you work with many boards, and many people do), so the best way is to described that in the boards' docs as included in this PR.

pfalcon commented:
>CM4 core does not starts by default, only CM0+ core works.

Ok, then why do we have cy8ckit_062_wifi_bt_m4 as part of this PR, if it doesn't really work (per Zephyr's convention, i.e. main() just runs). It definitely needs more work, so I'd suggest to move it to another (next) PR. But it would be really nice to make it work (again, by Zephyr conventions), as using UART with builtin USB-UART converter is very convenient (again, especially if someone works with many boards).

OlegKapshii commented:
@pfalcon 

> Either I misunderstood what you wrote, or that's not correct, according to CY8CKIT-062-WiFi-BT_Kit_Guide.pdf, section 4.3.4, page 54, which says that PSoC6's pins P5[0], P5[1], P5[2], P5[3] are converted to onboard KitProg2's USB-UART Bridge. I.e. M4's UART on P5 would be exposed in this way, not M0's on P12. I know it's easy to mix up such things (especially if you work with many boards, and many people do), so the best way is to described that in the boards' docs as included in this PR.

Yes, you are right. I do not have the board at home, so I looked into a schematic and misunderstood several pins. Now I see in the schematic that P12 port (which is used by CM0+) is connected to the socket pins. CM4 uses P5 pins which are translated to the USB.
Sorry. It is my mistake.
Sometimes there are bad contacts in sockets when you are using the UART-USB adapter, so check the wires. Check also that during connection to the COM port 115200, 8N1, none settings are used.

pfalcon commented:
@OlegKapshii 

> Sometimes there are bad contacts in sockets when you are using the UART-USB adapter, so check the wires. Check also that during connection to the COM port 115200, 8N1, none settings are used.

Sure, I keep all that stuff in mind. As I wrote, I have a suspicion that Linux' driver plays bad with Cypress USB-UART chip. If you want details, what I tried is flashing "scan" sample from WICED SDK, which output chunk of data over serial very second or so. It outputs over P5, so I normally see it using the board's builtin USB-UART converter. So what I did is connected this external USB-UART board you gave me to the same P5.0, P5.1 pins, and tried to compare the output which I get from /dev/ttyACM0 (board builtin) and /dev/ttyACM1 (external adapter). So, the builtin works fully as expected. Now with /dev/ttyACM1, I first of all get an extra empty line after each line of output, e.g. as if \r\n was translated to \r\n\n. And I get only very first chunk of output after I connect, then it's silent. And I can get more output though - for this, I need to press some keys, i.e. send some input into the adapter. That all sounds pretty crazy of course. I'm on kernel 4.13. Let me know if you ever tested that adapter with Linux, and what kernel it was.

The most annoying is that I couldn't find a spare USB-UART adapter - as anyone, I usually have a few lying around, but I dug thru all my hardware here (and that's a bunch), and couldn't find any %). Anyway, I ordered one, hope to get it tomorrow, and will test with it.

pfalcon commented:
> Now with /dev/ttyACM1, I first of all get an extra empty line after each line of output, e.g. as if \r\n was translated to \r\n\n. And I get only very first chunk of output after I connect, then it's silent.

Confirmed that with CP2102 based USB-UART, I don't get this behavior, the output is fully consistent and synchronous with KirProg2's output.

pfalcon commented:
Also confirming that with CP2102, I get serial output from the hello_world sample built for BOARD=cy8ckit_062_wifi_bt_m0 (on P12, as discussed above).

pfalcon commented:
@OlegKapshii, @nazrchorn : So, now need to just refactor the code as suggested above, extend docs, make CI pass, get people remove their -1's and give the +1's. Let me know if you don't agree with some suggestions and/or need help with them. Thanks!

carlescufi commented:
@OlegKapshii the approach is to always present the final layout of commits in the PR. So whenever you update your PR you rebase interactively and make the changes to the appropriate commits, or add/remove commits.

carlescufi commented:
@OlegKapshii that means of course that some of your commits, like the one named "Minor fixes after review" need to be squashed into existing commits.

pfalcon commented:
@OlegKapshii

> should all my previous commits, in some way, be reverted and recommitted in a right way (using rebase)?

Revert, no, these commits are the good base, they just need to be augmented/fixed up/cleaned up a bit. As mentioned by @carlescufi, the way to do it is git's interactive rebase feature (`git rebase --interactive` or `git rebase -i`).

I'm not sure how well you're versed with it, let me try to give a quick intro, but feel free to read man/google for it.

1. First of all, need to do non-interactive rebase to avoid surprises in next steps. This will also get rid of the merge commit. So, assuming you're on `add_psoc6` branch, run: `git rebase master`, then push the result: `git push -f`.
2. For starters, I'd suggest to perform interactive rebase on a separate branch, to be able to restart from scratch if something goes wrong. So, `git checkout -b add_psoc-rebase`.
3. Now just run `git rebase -i master`, and read carefully what will be written in the filed popping out in your editor. You should pay attention to: "If you remove a line here THAT COMMIT WILL BE LOST". The way I do and suggest to anyone is: perform single operation per interactive rebase interaction. I.e., if you need to make 10 changes, run git rebase -i ten times, verifying results after each - instead of trying to pack 10 changes into single run.
4. It's easy after that - e.g. to change a commit message, just replace default "pick" action with "reword" (or short "r"). E.g. when it was:

~~~
pick f6ecdc4267 ext: hal: cypress pdl: Integrate Cypress PDL into Zephyr Build system
~~~
change it to:

~~~
r f6ecdc4267 ext: hal: cypress pdl: Integrate Cypress PDL into Zephyr Build system
~~~

When you save change in your editor and quit, the actions will be performed, this commit message will be opened in your editor again, and you will be able to edit it.



Mastering "squash" or "fixup" operation will be next. Just keep in mind: squashing multiple commits together is much easier than splitting one commit into multiple. Don't haste with squashes, make sure you follow suggestions from the review above.

And you do have a commit which requires splitting - "Minor fixes after review". The way to do it is to use "edit" action, then when it stops on it:
1. git reset HEAD^
2. commit changes to each file individually (i.e. 3 commits).
3. continue rebase with git rebase --continue

Then start a new rebase -i to "fixup" those 3 newly created commits into initial commits which created corresponding files.


Let me know if you need other hints ;-). If you didn't practice rebase --interactive, it may seem daunting at first, but it's very powerful feature, and it's easy to master it (just follow suggestions like one change per rebase iteration). 80% of reviews require some changes, and that's what everyone doing in this case, so you're not alone! ;-)

pfalcon commented:
Another note, regarding Cypress HAL import:

~~~
zephyr/ext/hal/cypress/PDL/3.1.0$ du -sh *
4.0K	CMakeLists.txt
1.2M	cmsis
90M	devices
34M	doc
3.7M	drivers
28K	utilities
~~~

I doubt that we need doc/ folder in Zephyr. @galak, @nashif: FYI

OlegKapshii commented:
I did it. Thank you, @pfalcon . Your instructions were what I needed.

OlegKapshii commented on arch/arm/soc/cypress/Kconfig:
```diff
@@ -0,0 +1,34 @@
+#
+# Copyright (c) 2018, Cypress
```
I don't know what should it be here according to Linaro rules.

pfalcon replied:
> I don't know what should it be here according to Linaro rules.

Ok, as you submit this PR from @cypress.com emails, I guess copyright should be Cypress' too. Otherwise, deferring to @galak .

pfalcon commented:
@OlegKapshii : Thanks for going thru that. I'd say, commits look much better now, but I didn't look into them in detail yet, because there's another problem: this PR now contains 19 commits, whereas previously it was 11.

It looks like the new changes were git-merged into the old PR, so it now contains both old and new sets. That's not the desired situation, the new changes were expected to replace the old ones completely. Generally, merge-based workflow is not compatible with rebase-based workflow, so either one or another should be used, and Zephyr uses rebase-based workflow. That means for example that "git pull" should not be used, instead "git pull --rebase" should be.


Anyway, let me look more into this and come up with suggestions on how to fix it.

pfalcon commented:
Ok, here's how I was able to fix it. First of all, the recommended setup is that "master" branch always tracks upstream Zephyr and is up to date. I have "upstream" remote for upstream repo:

~~~
$ git remote get-url upstream
https://github.com/zephyrproject-rtos/zephyr
~~~

So, making sure master is up to date:

~~~
git checkout master
git pull --rebase upstream master
~~~

Now on to fix the `add_psoc6` branch:

~~~
git checkout add_psoc6
git rebase -i master
~~~

In editor, I see (empty line inserted by me):

~~~
pick 1c42c4056e ext: hal: cypress: Import Cypress PDL for PSoC6 SoC Support
pick 84bbb34e4a ext: hal: cypress: Integrate Cypress PDL into Zephyr Build system
pick 72160e4f5b ext: hal: cypress: Protect floating point code with CONFIG_FLOAT
pick 3b1077585d arm: Add Cypress PSoC6 SoC support
pick 7daf0546c4 drivers: serial: Add Cypress PSoC6 UART driver
pick c9398fb170 boards: arm: Add Cypress PSoC6 cy8ckit_062_wifi_bt_m0 board support
pick d91de0e1ac boards: arm: Add Cypress PSoC6 cy8ckit_062_wifi_bt_m4 board support
pick 3c8c3fbf5a drivers: serial: Refactored Cypress PSoC6 UART driver

pick f6ecdc4267 ext: hal: cypress pdl: Integrate Cypress PDL into Zephyr Build system
pick d38c686db5 ext: cypress PDL: Protect floating point code with CONFIG_FLOAT
pick c0d977f666 arm: add PSOC6 SoC support
pick d904b7af1e add psoc6 uart driver
pick 548ebbf7f7 Add board support for cy8ckit_062_wifi_bt_m0
pick 4d298b6965 add boards cy8ckit_062_wifi_bt_m4
pick 259d5c21c3 Minor fixes after review
pick 826fbc7618 drivers: serial: Refactored Cypress PSoC6 UART driver
pick 75c00b377e Refactored PSoC6 UART driver
~~~


So, what we need is first set of commits. 2nd set after the empty line are old duplicates, it should be gone (@OlegKapshii, please verify that nothing is omitted in 1st chunk of course).

So, we leave in the editor only:

~~~
pick 1c42c4056e ext: hal: cypress: Import Cypress PDL for PSoC6 SoC Support
pick 84bbb34e4a ext: hal: cypress: Integrate Cypress PDL into Zephyr Build system
pick 72160e4f5b ext: hal: cypress: Protect floating point code with CONFIG_FLOAT
pick 3b1077585d arm: Add Cypress PSoC6 SoC support
pick 7daf0546c4 drivers: serial: Add Cypress PSoC6 UART driver
pick c9398fb170 boards: arm: Add Cypress PSoC6 cy8ckit_062_wifi_bt_m0 board support
pick d91de0e1ac boards: arm: Add Cypress PSoC6 cy8ckit_062_wifi_bt_m4 board support
pick 3c8c3fbf5a drivers: serial: Refactored Cypress PSoC6 UART driver
~~~

and save.

Verifying that everything builds for BOARD=cy8ckit_062_wifi_bt_m0 and BOARD=cy8ckit_062_wifi_bt_m4. And force push into this PR:

~~~
git push -f
~~~

pfalcon commented:
In the future to pull changes from master branch, use:

~~~
git checkout master
git pull --rebase master
git checkout add_psoc6
git rebase master
git push -f
~~~

OlegKapshii commented:
It seems, fixed

pfalcon commented:
@OlegKapshii : Cool, thanks!

OlegKapshii commented:
@pfalcon 
There is a number of style-errors according to Shippable tests. Should I fix it to finish the PR?

pfalcon commented on boards/arm/cy8ckit_062_wifi_bt_m4/cy8ckit_062_wifi_bt_m4.yaml:
```diff
@@ -0,0 +1,9 @@
+name: Cypress PSoC6 WiFi-BT Pioneer Kit
+type: mcu
+arch: arm
+ram: 96
```
Isn't there 288K of RAM?

pfalcon commented on boards/arm/cy8ckit_062_wifi_bt_m4/cy8ckit_062_wifi_bt_m4.yaml:
```diff
@@ -0,0 +1,9 @@
+type: mcu
+arch: arm
+ram: 96
+flash: 512
```
Isn't there 1024K of Flash?

pfalcon commented on arch/arm/soc/cypress/psoc6/soc.c:
```diff
@@ -0,0 +1,387 @@
+}
+
+/*******************************************************************************
+* Function Name: InitHardware()
```
Stray comment block, should be removed.

pfalcon reviewed (commented):
Few more comments.

OlegKapshii commented:
@pfalcon 
I have splitted/squashed the  "drivers: serial: Refactored Cypress PSoC6 UART driver" commit. All the code works fine on my PC, but I see some built errors in the Shippable log. Could you point me where the problem is, please?

pfalcon commented:
> I have splitted/squashed the "drivers: serial: Refactored Cypress PSoC6 UART driver" commit. 

Nice, thanks! Now next step is to finally rebase in on main Zephyr master. If you look at https://github.com/nazrchorn/zephyr/tree/add_psoc6 , it says "This branch is 7 commits ahead, 1267 commits behind zephyrproject-rtos:master.", so this branch is quite a lot behind the current state of development. When you rebase, you'll notice that https://github.com/zephyrproject-rtos/zephyr/pull/9776 was merged now, which moved `soc/` directories from under `arch` to the top-level one. 

So, what was `arch/arm/soc/cypress/CMakeLists.txt`, now needs to be `soc/arm/cypress/CMakeLists.txt`. Now that the commits are clean, it should be easy to do: you just look what files were introduce in a particular commit, in your branch do  `git mv ...` just for those files, commit as a fix-up commit, then via rebase squash into the main commit. Repeat for other commits as needed.
 

> All the code works fine on my PC, but I see some built errors in the Shippable log. Could you point me where the problem is, please?

So, steps above which would need to be done so far. Once the changes are on the top of latest master, we can look into making final pass to fix any CI issues, I'll be looking into them, yes.

Thanks!

OlegKapshii commented on arch/arm/soc/cypress/Kconfig:
```diff
@@ -0,0 +1,34 @@
+
+config SOC_FAMILY_PSOC6
+	bool
+	# omit prompt to signify a "hidden" option
```
Not sure that I understand correctly - should be config SOC_FAMILY_PSOC6... moved before the SOC_PSOC6_M0 and SOC_PSOC6_M4?

pfalcon replied:
Sorry, there was a typo in my comment - "commit" used instead of "comment".

And I grepped the exiting code, it has those lines in this order. So, sorry for the noise, nothing needs to be done.

OlegKapshii commented:
@pfalcon 
So, I have finished work in this PR and there are even no failed tests in Shippable.
Should I do here something else?

OlegKapshii commented:
@pfalcon 
And now there is a question how can I add PSoC6 IPM related code in Zephyr? As far as I understand - I cannot make a fork from the add_psoc6 branch and cannot commit in the add_psoc6 branch more. I cannot use also the master branch to fork from it, because it does not contain PSoC6 related code. What should I do here? Can you suggest anything, please?
Thank you

pfalcon commented on dts/arm/cypress/psoc6.dtsi:
```diff
@@ -0,0 +1,83 @@
+
+	flash-controller@4001E000 {
+			compatible = "nrf,nrf51-flash-controller";
+			reg = <0x4001E000 0x518>;
```
As mentioned in previous comment, these look like values for Nordic nrf5xxxx chip.

pfalcon commented on dts/arm/cypress/psoc6.dtsi:
```diff
@@ -0,0 +1,83 @@
+
+			flash0: flash@10000000 {
+				compatible = "soc-nv-flash";
+				label = "NRF_FLASH";
```
FLASH_0

pfalcon commented on dts/arm/cypress/psoc6.dtsi:
```diff
@@ -0,0 +1,83 @@
+
+			flash1: flash@10060000 {
+				compatible = "soc-nv-flash";
+				label = "NRF_FLASH";
```
FLASH_1

pfalcon reviewed (commented):
@OlegKapshii, found some issues with Flash controller definition still, should be easy to fix.

pfalcon commented:
@OlegKapshii:

> So, I have finished work in this PR and there are even no failed tests in Shippable.

Great, thanks!

> Should I do here something else?

I had a look, and there're still some references to "nrf" flash controller still. Can you please fix those?

Then we should ping folks who previously left -1 review asking them to re-review. 

@dbkinder, Can you have a look at the docs, they should be updated based on your earlier comments.

pfalcon commented:
> And now there is a question how can I add PSoC6 IPM related code in Zephyr? As far as I understand - I cannot make a fork from the add_psoc6 branch and cannot commit in the add_psoc6 branch more. I cannot use also the master branch to fork from it, because it does not contain PSoC6 related code. What should I do here? Can you suggest anything, please?

Yes, please keep this branch just for the initial addition and this PR. It may receive a fixup/cleanup commits yet (to be squashed), but no new functionality should be added indeed, to not put off review/merge further.

But what to do is simple: just create a new branch based on this branch, and work on it. It thus will depend on "add_psoc6" instead of master, and that's fine. 

If you have "add_psoc6" branch active, just do `git checkout -b psoc6-ipm`, or whatever you want to call it.

The whole dependency picture will be:

upstream master -> master -> add_psoc6 -> psoc6_ipm

The only trick is that you should keep rebase workflow for all these branches, i.e. rebase one on top of another in cascade. I guess, if you just create psoc6_ipm now (or next week), you might work w/o rebasing for some time, so we can discuss it later.

OlegKapshii commented:
Fixed all errors. All shippable tests passed.

pfalcon reviewed (approved):
@OlegKapshii : Thanks for all the refactoring and cleanup! Looks good to me, approving.

pfalcon commented:
@cvinayak : Your -1 was based on references to "nrf" in some of DTS files, those were fixed now, can you please re-review?

@galak: Can you please have a look too?

pfalcon commented:
@dbkinder : Friendly ping about docs. Thanks!

galak commented on dts/arm/cypress/mem.h:
```diff
@@ -0,0 +1,34 @@
+
+#include <mem.h>
+
+#define DT_FLASH_ADDR		0x10000000
```
lets remove this file and just merge these into the .dtsi files.

dbkinder reviewed (changes requested):
documentation has important sections with no content (TBD)

pfalcon commented on boards/arm/cy8ckit_062_wifi_bt_m4/doc/cy8ckit_062_wifi_bt.rst:
```diff
@@ -0,0 +1,118 @@
+Interrupt Controller
+====================
+
+TBD
```
@dbkinder: We'll definitely need to/will elaborate docs going forward, it's a typical situation that we can't get everything into perfect shape from the start.

@OlegKapshii, @dbkinder : But here I agree that TBD here is not much good, but my suggestion is to just remove "Interrupt Controller" section altogether. Looking at doc files for existing boards, only few have this section. Indeed, that's hardly an important section for *Zephyr RTOS* docs. The RTOS handles interrupts, that's all user should know. More details can be founds in datasheet/TRM for a particular SoC.

dbkinder replied:
I'll be disappointed with empty TBD sections, but won't stop the PR :(

pfalcon commented on boards/arm/cy8ckit_062_wifi_bt_m0/doc/cy8ckit_062_wifi_bt_m0.rst:
```diff
@@ -0,0 +1,106 @@
+Interrupt Controller
+====================
+
+TBD
```
@OlegKapshii: As above, please remove the whole section for now.

galak reviewed (changes requested):
Add copyright/license to files like CMakeLists.txt

galak commented on boards/arm/cy8ckit_062_wifi_bt_m0/CMakeLists.txt:
```diff
@@ -0,0 +1,3 @@
+zephyr_library()
```
Add copyright/license header

galak commented on boards/arm/cy8ckit_062_wifi_bt_m0/board.cmake:
```diff
@@ -0,0 +1,3 @@
+include(${ZEPHYR_BASE}/boards/common/bossac.board.cmake)
```
Why are we using/setting bossac here?

galak commented on boards/arm/cy8ckit_062_wifi_bt_m0/board.h:
```diff
@@ -0,0 +1,12 @@
+/*
+ * Copyright (c) 2015-2016 Intel Corporation
```
Fix copyright, shouldn't be Intel or 2015-2017

galak commented on boards/arm/cy8ckit_062_wifi_bt_m0/cy8ckit_062_wifi_bt_m0.dts:
```diff
@@ -0,0 +1,23 @@
+/dts-v1/;
```
Add copyright/license header

galak commented on boards/arm/cy8ckit_062_wifi_bt_m0/cy8ckit_062_wifi_bt_m0_defconfig:
```diff
@@ -0,0 +1,25 @@
+CONFIG_ARM=y
```
Add copyright/license header

galak commented on boards/arm/cy8ckit_062_wifi_bt_m0/pinmux.c:
```diff
@@ -0,0 +1,33 @@
+/* pinmux_board_arduino_due.c - Arduino Due pinmux driver */
```
remove arduino reference

galak commented on boards/arm/cy8ckit_062_wifi_bt_m0/pinmux.c:
```diff
@@ -0,0 +1,33 @@
+/* pinmux_board_arduino_due.c - Arduino Due pinmux driver */
+
+/*
```
Fix copyright  / year

galak commented on boards/arm/cy8ckit_062_wifi_bt_m0/pinmux.c:
```diff
@@ -0,0 +1,33 @@
+
+#include "pinmux/pinmux.h"
+
+static void __pinmux_defaults(void)
```
remove __pinmux_defaults for now

galak commented on boards/arm/cy8ckit_062_wifi_bt_m0/support/openocd.cfg:
```diff
@@ -0,0 +1,4 @@
+
```
Does this work?  If not lets remove this file for now.

galak reviewed (changes requested):
A number of cleanups for the m0 board support additions

galak reviewed (changes requested):
Various feedback on the m4 board port changes

galak commented on dts/arm/cypress/psoc6.dtsi:
```diff
@@ -0,0 +1,79 @@
+#include <mem.h>
```
Add missing copyright / license

galak commented on dts/arm/cypress/psoc6_cm0.dtsi:
```diff
@@ -0,0 +1,16 @@
+#include <arm/armv6-m.dtsi>
```
Add missing copyright / license

galak commented on dts/arm/cypress/psoc6_cm4.dtsi:
```diff
@@ -0,0 +1,16 @@
+#include <arm/armv7-m.dtsi>
```
Add missing copyright / license

galak commented on dts/bindings/flash_controller/cypress,psoc6-flash-controller.yaml:
```diff
@@ -0,0 +1,16 @@
+---
```
Add missing copyright / license

galak commented on soc/arm/cypress/Kconfig:
```diff
@@ -0,0 +1,33 @@
+
+config SOC_PSOC6_M4
+	bool "SOC_PSOC6_M4"
+	select CPU_CORTEX_M4
```
Should we select CPU_HAS_FPU

OlegKapshii replied:
When I add "select CPU_HAS_FPU" after "select CPU_CORTEX_M4" it turns on CONFIG_FLOAT and the problem with math.h include appears.
What is better to do here?

galak replied:
Lets leave it alone for now, and I'll take a look after we get this merged

galak commented on soc/arm/cypress/psoc6/soc.c:
```diff
@@ -0,0 +1,387 @@
+	.cco_Freq = 355u,
+};
+
+__STATIC_INLINE void Cy_SysClk_ClkFastInit(void)
```
replace all __STATIC_INLINE with static inline

galak commented on soc/arm/cypress/psoc6/soc.c:
```diff
@@ -0,0 +1,387 @@
+}
+
+
+void init_cycfg_platform(void)
```
static void init_cycfg_platform(void)?

galak commented on soc/arm/cypress/psoc6/soc.c:
```diff
@@ -0,0 +1,387 @@
+/**
+ * Function Name: InitHardware()
+ *
+ *  Peforms hardware initialization: LEDs.
```
Fix this comment block.  Function name doesnt match, reference to LEDs doesn't make sense.

galak reviewed (changes requested):
Various comments on basic SoC changes for PSOC6

galak commented on dts/bindings/serial/cypress,psoc6-uart.yaml:
```diff
@@ -0,0 +1,27 @@
+---
```
Add missing copyright / license info

galak reviewed (changes requested):
Add missing copyright/license

OlegKapshii commented on boards/arm/cy8ckit_062_wifi_bt_m0/doc/cy8ckit_062_wifi_bt_m0.rst:
```diff
@@ -0,0 +1,106 @@
+.. _cy8ckit_062_wifi_bt_m0:
+
+PSoC6 WiFi-BT Pioneer Kit
```
@dbkinder 
Sorry, but I do not understand your point. Can you provide me more details, please?
It is made similarly to lpcxpresso54114_m0 and lpcxpresso54114_m4 boards.

dbkinder replied:
The recommendation is to change the title of these documents to be:   ``PSoC6 WiFi-BT Pioneer Kit (M0)`` and ``PSoC6 WiFi-BT Pioneer Kit (M4)``, as was done for the lpcxpresso54114 boards.  Otherwise, when you look at the table of contents (https://docs.zephyrproject.org/latest/boards/boards.html) the two boards will show the same title.

OlegKapshii replied:
Now I see. Thank you

OlegKapshii commented:
Fixed

galak commented:
@dbkinder how do you feel about the docs?

dbkinder reviewed (approved):
I'm sad there are no build, flash, or run instructions, but I promised not to hold up the PR because of that.  Please add those instructions soon.  Otherwise, LGTM.

OlegKapshii commented:
No problems with the build instructions and example of hello_world run/output. I can add it simply.

The problem could with the flash instruction. I am using our internal tools to write elf file. I need to talk with one of my team-mates to get some good instruction. I believe, that I will have it on Monday.
So, if you wait I will add this information on Monday.

galak commented:
> No problems with the build instructions and example of hello_world run/output. I can add it simply.
> 
> The problem could with the flash instruction. I am using our internal tools to write elf file. I need to talk with one of my team-mates to get some good instruction. I believe, that I will have it on Monday.
> So, if you wait I will add this information on Monday.

We might merge this and can cleanup the docs as a new PR on Monday if you get the info.

We're in the process of possibly updated the OpenOCD version that is in the Zephyr SDK to the top of tree for openocd.  It looks like there is some support for the PSoC6 that got merged, so will be interesting if we can flash with it.
